### PR TITLE
PBES2 fix

### LINKED ISSUE #50: TypeError: str.charCodeAt is not a function
I try to run the application, but i receive this error. any solution ?

### PR DESCRIPTION
**Pull request by cryptism**
Fixes #50 

Later versions of Firefox use PBES2 encryption, which changed the certificate structure and decryption mechanism

### CODE REVIEW

File: src/renderer.js [hunk 1]
```diff
@@ -219,68 +214,35 @@ const iconv = require('iconv-lite');
...
-        const decipher = forge.cipher.createDecipher('3DES-CBC', key);
+    // Adapted from https://github.com/lclevy/firepwd/blob/master/firepwd.py
+    function decryptPbe(decodedItemSeq, password, globalSalt) {
+        const data = decodedItemSeq[1].value;
+        const pbkdf2Seq = decodedItemSeq[0].value[1].value[0].value[1].value;
+        const salt = pbkdf2Seq[0].value;
+        const iterations = pbkdf2Seq[1].value.charCodeAt();
+        const keyLength = pbkdf2Seq[2].value;
+        // Prepending 0x040e, where 0x04 = octetstring, 0x0e = string length (14)
+        const iv = '' + decodedItemSeq[0].value[1].value[1].value[1].value;
```
Comment (cryptism) [inline src/renderer.js:225]: Looks like a bug on Github's end. Copy this into some text and you should see that these are characters that don't have a Unicode representation.

File: package.json [hunk 2]
```diff
@@ -91,9 +91,9 @@
     "devDependencies": {
         "electron": "^3.0.6",
         "electron-builder": "^20.29.0",
-        "electron-reload": "^1.2.5",
+        "electron-reload": "^1.5.0",
```
Comment (kspearrin) [inline package.json:94]: Was there a reason this was updated?

File: package.json [hunk 3]
```diff
@@ -91,9 +91,9 @@
...
         "eslint": "^5.8.0",
-        "node-forge": "^0.7.6",
+        "node-forge": "^0.9.1",
```
Comment (kspearrin) [inline package.json:96]: Also, wondering why this was updated? Was there some fix/addition needed?

File: package.json (hunk 2 above)
Comment (cryptism) [reply package.json:94]: `npm audit` ran a bunch of critical package updates, so I ran `fix` on them. More a force of habit than anything

**Comment by kspearrin** [comment]
@cryptism Just testing this on my FF 77.0 on Windows 10. I get the following error:

```
TypeError: Cannot read property '0' of undefined
    at decryptPbe (C:\Projects\ff-password-exporter\src\renderer.js:221:31)
    at getKey (C:\Projects\ff-password-exporter\src\renderer.js:200:32)
    at getLogins (C:\Projects\ff-password-exporter\src\renderer.js:135:21)
    at HTMLButtonElement.exportBtn.addEventListener (C:\Projects\ff-password-exporter\src\renderer.js:58:28)
```

It comes from this line:

```
const pbkdf2Seq = decodedItemSeq[0].value[1].value[0].value[1].value;
const salt = pbkdf2Seq[0].value;
```

`pbkdf2Seq` appears to be undefined for me. I don't have any passwords saved in FF, or do I have a master password set. So maybe that is why? We might want to gracefully handle these errors cases better?

**Comment by kspearrin** [comment]
I saved a login in Firefox and was still getting the same error message. I then added a master password and now I get a new error:

```
abort("Cannot enlarge memory arrays. Either (1) compile with  -s TOTAL_MEMORY=X  with X higher than the current value 16777216, (2) compile with  -s ALLOW_MEMORY_GROWTH=1  which allows increasing the size at runtime but prevents some optimizations, (3) set Module.TOTAL_MEMORY to a higher value before the program runs, or (4) if you want malloc to return NULL (0) instead of this abort, compile with  -s ABORTING_MALLOC=0 ") at Error
    at jsStackTrace (C:\Projects\ff-password-exporter\node_modules\sql.js\js\sql.js:3:13718)
    at stackTrace (C:\Projects\ff-password-exporter\node_modules\sql.js\js\sql.js:3:13889)
    at abort (C:\Projects\ff-password-exporter\node_modules\sql.js\js\sql.js:20:8469)
    at abortOnCannotGrowMemory (C:\Projects\ff-password-exporter\node_modules\sql.js\js\sql.js:3:14751)
    at enlargeMemory (C:\Projects\ff-password-exporter\node_modules\sql.js\js\sql.js:3:15210)
    at Wt (C:\Projects\ff-password-exporter\node_modules\sql.js\js\sql.js:12:208484)
    at bs (C:\Projects\ff-password-exporter\node_modules\sql.js\js\sql.js:12:168983)
    at Gf (C:\Projects\ff-password-exporter\node_modules\sql.js\js\sql.js:7:216557)
    at fp (C:\Projects\ff-password-exporter\node_modules\sql.js\js\sql.js:10:147138)
    at ep (C:\Projects\ff-password-exporter\node_modules\sql.js\js\sql.js:10:146233)
If this abort() is unexpected, build with -s ASSERTIONS=1 which can give more information.
```

**Comment by kspearrin** [comment]
@cryptism Nevermind on the memory error for SQL.js. That seemed to be a node runtime issue that was resolved in release builds.

I got it working and exporting my passwords after running a release build.

The original "Cannot read property '0' of undefined" error seems to still occur on profiles that do not have a master password. We need to figure out how to handle that case.

**Comment by cryptism** [comment]
> @cryptism Nevermind on the memory error for SQL.js. That seemed to be a node runtime issue that was resolved in release builds.
> 
> I got it working and exporting my passwords after running a release build.
> 
> The original "Cannot read property '0' of undefined" error seems to still occur on profiles that do not have a master password. We need to figure out how to handle that case.

Ah, thanks for bringing this up. I've successfully imported my firefox passwords over to bitwarden now, so I'm happy to work on this edge case.

**Comment by cryptism** [comment]
@kspearrin Hmm, so I removed my master password and it's still working for me. Are we definitely using version of Firefox (77.01)?

**Comment by cryptism** [comment]
Purely speculative but it's possible this could still be using the old cert encryption, which possibly should have been kept in for compatibility's sake.

**Comment by mnoorenberghe** [comment]
The iteration count only gets upgraded to 10,000 if you set a master password after the improvement was made in NSS/Firefox. You probably need to handle that. You can test with a new Firefox profile from [ESR 60](https://archive.mozilla.org/pub/firefox/releases/60.9.0esr/) which is before the key derivation was made stronger.

**Comment by cryptism** [comment]
@mnoorenberghe Awesome, thanks. I can now repeat the issue.

**Comment by cryptism** [comment]
We're supporting an arbitrary iteration count, I think it just looks like this issue is due to us not handling the old PBES scheme. #52 brings it back, and now all the profiles work for me

**Pull request merged**